graphql_compiler/schema_merging/rename_schema.py

```diff
@@ -0,0 +1,232 @@
+        })
+        self.rename_types = frozenset({
+            'NamedType', 'ObjectTypeDefinition', 'InterfaceTypeDefinition',
+            'UnionTypeDefinition', 'EnumTypeDefinition'
+        })
+        self.unexpected_types = frozenset({
+            'OperationDefinition', 'SelectionSet', 'Field', 'FragmentSpread',
+            'InlineFragment', 'FragmentDefinition'
+        })
+
+    def _rename_name_add_to_record(self, node):
+        """Rename the value of the node, and add the name mapping to reverse_name_map.
+
+        Don't rename if the type is the query type, a scalar type, or a builtin type.
+
+        Args:
+            node: Name type Node
+
+        Raises:
+            SchemaError if the newly renamed node causes name conflicts with existing types,
+                scalars, or builtin types
+        """
+        name_string = node.value
+
+        if (
+            name_string == self.query_type or
+            name_string in self.scalar_types or
+            name_string in self.builtin_types
+        ):
+            return
+
+        new_name_string = self.rename_func(name_string)
+
+        if (
+            new_name_string in self.reverse_name_map and
+            self.reverse_name_map[new_name_string] != name_string
+        ):
+            raise SchemaError(
+                '"{}" and "{}" are both renamed to "{}"'.format(
+                    name_string, self.reverse_name_map[new_name_string], new_name_string
+                )
+            )
+        if new_name_string in self.scalar_types:
+            raise SchemaError(
+                '"{}" was renamed to "{}", clashing with scalar "{}"'.format(
+                    name_string, new_name_string, new_name_string
+                )
+            )
+        if new_name_string in self.builtin_types:
+            raise SchemaError(
+                '"{}" was renamed to "{}", clashing with builtin "{}"'.format(
+                    name_string, new_name_string, new_name_string
+                )
+            )
+
+        node.value = new_name_string
+        self.reverse_name_map[new_name_string] = name_string
+
+    def enter(self, node, key, parent, path, ancestors):
+        node_type = type(node).__name__
+        if node_type in self.noop_types:
+            # Do nothing, continue traversal
+            return None
+        elif node_type in self.rename_types:
+            # Rename and put into record the name attribute of current node; continue traversal
+            self._rename_name_add_to_record(node.name)
+        elif node_type in self.unexpected_types:
+            # Node type unexpected in schema definition, raise error
+            raise SchemaError('Node type "{}" unexpected in schema AST'.format(node_type))
+        elif node_type == 'TypeExtensionDefinition':
+            raise SchemaError('Extension definition not allowed')
+        else:
+            # VariableDefinition, Variable, ObjectValue, ObjectField, InputObjectTypeDefinition
```

Comment:
We don't use `InputObjectType` objects in the schema so we can ignore that. I think that`Variable` and `VariableDefinition` are only relevant for queries since I've never seen the `Variable` in schema generation. Not sure about `ObjectValue` but based on the name `ObjectField` might be referring to the field of `GraphQLObjectType` objects so I would double-check that it isn't important.

Reply:
So `InputObjectType` can exist in schemas, but we don't use it/allow it in inputs? And `Variable` and `VariableDefinition` cannot exist in schemas? I'll try to figure out use of ObjectField and ObjectValue more.

Reply:
From what I can tell, it seems like `ObjectValue` and `ObjectField` only appear in things like query results. Schemas just have `FieldDefinition` and `ObjectTypeDefinition`. For now I'll categorize those as types that are unexpected in schemas. If this turns out to be wrong, we can fix it.

Reply:
From what I can tell, ignoring these 5 types for now seems to be the right thing to do and if it isn't we can fix it later. Furthermore, if it isn't the right to do, then we should be easily be able to tell from the tests.